Philips TV ambilight support

elupus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
* Add support for ambilight control

Bump of dependency: https://github.com/danielperna84/ha-philipsjs/compare/2.7.3...2.7.4

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16778

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

elupus commented:
Should be done for start of review. I still need to update docs.

joncar commented on homeassistant/components/philips_js/__init__.py:
```diff
@@ -1 +1,97 @@
+
+
+class PhilipsTVDataUpdateCoordinator(DataUpdateCoordinator[None]):
+    """Define a wrapper class to update Awair data."""
```
"Awair"? Copy and paste leftover? Is also in the comment on __init__.

joncar commented on homeassistant/components/philips_js/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
```
Comment is out of date: STEP_USER_DATA_SCHEMA isn't a defined constant. Instead you've defined it inline later in this file.

joncar commented on homeassistant/components/philips_js/device_trigger.py:
```diff
@@ -0,0 +1,69 @@
+"""Provides device automations for Arcam FMJ Receiver control."""
```
Copy and pasted comment needs updating

joncar commented on homeassistant/components/philips_js/device_trigger.py:
```diff
@@ -0,0 +1,69 @@
+
+
+async def async_get_triggers(hass: HomeAssistant, device_id: str) -> List[dict]:
+    """List device triggers for Arcam FMJ Receiver control devices."""
```
Copy and pasted comment needs updating

joncar commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,174 @@
+"""Media Player component to integrate TVs exposing the Joint Space API."""
```
Copy and pasted comment needs updating

joncar commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,174 @@
+        effect = kwargs.get(ATTR_EFFECT)
+
+        if not self._tv.on:
+            raise Exception("Light is not turned on")
```
This is a bit unusual. Would it be appropriate to indicate more clearly why this doesn't work and what should be done instead? I'm guessing its something like "Amblight can only be set if the TV is on."?

elupus replied:
Well yes. The TV can not be turned on via the API. We provide an event on the media player to hook up an ir blaster  to turn on. But as the light is currently set up. It's always on if the TV is turned on. It's the EFFECT that actually controls how the light behaves.

elupus replied:
I've reworded it.

joncar commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,174 @@
+        self._update_from_coordinator()
+        super()._handle_coordinator_update()
+
+    def turn_on(self, **kwargs) -> None:
```
Should there be an implementation of turn_off? Perhaps it could set the color to black?

joncar commented on homeassistant/components/philips_js/media_player.py:
```diff
@@ -192,45 +175,44 @@ def is_volume_muted(self):
-        if self._on_script:
-            self._on_script.run(context=self._context)
-            self._update_soon(DELAY_ACTION_ON)
+        _LOGGER.debug("Firing event to turn on device")
```
Is this debug line still needed?

elupus replied:
Well maybe not. It's hard to understand what is going on if you only are looking at logs for this component.

joncar commented on tests/components/philips_js/test_device_trigger.py:
```diff
@@ -0,0 +1,97 @@
+"""The tests for Arcam FMJ Receiver control device triggers."""
```
Copy and pasted comment

joncar commented on tests/components/philips_js/test_device_trigger.py:
```diff
@@ -0,0 +1,97 @@
+
+
+async def test_get_triggers(hass, device_reg, entity_reg):
+    """Test we get the expected triggers from a arcam_fmj."""
```
Copy and pasted comment

elupus committed:
Add ambilight support

joncar reviewed (approved):
Looks good to me!

Quentin-Cloos commented:
I've just tried it and it's working great for now! Like others Ambilight component it just can't turn on the light when the TV is in deep sleep (or am I missing something?) but with Wakelock app it's working. I have tried transitions but it's not working too, the color changing is instant. Anyway it's great to see an official component for our Ambilight TV! It's just missing a switch for the Ambilight+Hue, I've made one myself but I'm not really a dev. Sorry for my approximate english and thank you again for your work!

chpego commented:
@elupus thank you very much for your work. i've tried and for me it's works for only color, but i can't see effect video on the list and, i don't know why. Do you need something to debug ? I'm using a Philips 55oled804/12 with api 6 version

elupus commented:
At the moment it is using the old ambilight control api that is available for v1. Annoyingly they had to go and invent new stuff for v6.

I will likely add support for those styles in a later pull.

chpego commented:
Thank you @elupus ... i mentionned an issue on the other github project, i've made some modification, maybe you can (re)use to include on ha-philipsjs ?

elupus commented:
ha-philipsjs already have support for getting styles. Just not used here yet. It need to fully replace the old logic if the system support it. 

Ps. I don't know which other project you mean.

Quentin-Cloos commented:
I have found another "issue", I don't know if it's a normal behavior but when I change the color to pure red (255, 0, 0), the color on the Ambilight is orange, I can live with it but I prefer to share this with you 🙂

elupus commented:
> I have found another "issue", I don't know if it's a normal behavior but when I change the color to pure red (255, 0, 0), the color on the Ambilight is orange, I can live with it but I prefer to share this with you 🙂

That is really odd. Does it snap to orange? That is, something really close to pure orange is still red:er?

Quentin-Cloos commented:
> That is really odd. Does it snap to orange? That is, something really close to pure orange is still red:er?

The orange is more yellow, when I test the Ambilight in follow color mode I have a real red, but I forgot to say that I always had this issue with Ambilight custom components, the colors are not the same comparing to my Philips bulbs. Am I the only one?

aegjoyce commented:
Hi elupus, this is some fantastic work! Integration and setup works brilliantly. Love the improved functionality over the original android_tv component.

I haven't experienced the same colour replication issues as the commenter above but would suggest that colour replication can be skewed if the TV hasn't had ambilight calibration from an ISF setting.

I am having the same problem with the ambilight not turning off properly when the TV is off - mine appears to get set to the follow video mode after initially turning off. Might try playing around with light.py and your previous commits to see if I can get this working

elupus commented:
Yea that not returning to standard is a bug on mine too. No idea on how to fix that. I likely need to get some known issues in the docs on stuff like this.

aegjoyce commented:
Bearing in mind I have no idea what I'm doing and am just feeling my way around the code - this seems to work reasonably well:

```
    async def async_turn_off(self, **kwargs) -> None:
        """Turn off ambilight."""

        if not self._tv.on:
            raise Exception("TV is not available")
            
        if self._tv.ambilight_power == "On":
            if not await self._tv.setAmbilightPower("Off"):
                raise Exception("Failed to set ambilight power")

        # current = self._tv.ambilight_current_configuration
        # if current and current["isExpert"]:
        #     config: AmbilightCurrentConfiguration = {
        #         "styleName": "FOLLOW_VIDEO",
        #         "isExpert": False,
        #         "menuSetting": "Standard",
        #     }
        #     if not await self._tv.setAmbilightCurrentConfiguration(config):
        #         raise Exception("Failed to set ambilight configuration")
        # else:
        #     if not await self._tv.setAmbilightMode("internal"):
        #         raise Exception("Failed to set ambilight mode")

        self._update_from_coordinator()
        self.async_write_ha_state()
```

There's a bit of a delay to update status when turning off from any mode other than Manual, but it seems to do the trick...

elupus commented:
That would turn off ambilight even after you turn on your tv again thou. At least that is how it works on my tv.

elupus commented:
Hmm. But maybe "Have you tried turning it off and on again" like classical IT support :)

aegjoyce commented:
> That would turn off ambilight even after you turn on your tv again thou. At least that is how it works on my tv.

Mine seems to work as expected when turning back on after, no issues. If it's device-specific though you could include something in media_player.py that turns ambilight on when the tv is switched on?

Quentin-Cloos commented:
I have a major issue but I'm not sure it's related with your component, sometime the Philips TV become unavailable from Home Assistant and I can't control the Ambilight anymore, the only fix I found is to restart the TV then to restart HA. But I noticed that when it happened I can't control my TV anymore with the TV Remote app nor with https rest commands (but http rest commands still work, like http://IP:1925/6/ambilight/currentconfiguration). Is it possible that your component is flooding the JointSpace API (just a guess) and that the API stop responding because of that? The main issue is that all my automation that have the Ambilight in it stop working (because HA gave me errors). Sorry again for my frenglish and thank you again for your help and your time! 🙂

elupus commented:
Interesting that it still is accessable over http. Sound like some hanging connection on http. I allow two https connections maximum. I need more than one since the notify feature hogs one all the time. I can't cancel the notify one since it would still be hanging on the tv.

I wonder if the notify feature works over standard http. That might be an option to avoid more connections.

elupus commented:
Btw. This hang/crash only happens when screen is turned off right? That is when I have ended up in this situation.

Quentin-Cloos commented:
> Btw. This hang/crash only happens when screen is turned off right? That is when I have ended up in this situation.

I have this issue when the screen is on too, it happened when I'm using the TV for no reason.

Quentin-Cloos commented:
Is it possible that this issue is caused by the fact that I'm still using my old login and password for sending rest commands to JointSpace from Home Assistant? This is the only way I found to enable Ambilight+Hue and to put the Follow_Video mode. Is there a way to know what are the new login and password from the new pairing process? Or can you had this modes to the light.py or maybe a switch.py for the Ambilight+Hue, with that your component will be perfect 😃 I still have a lot of "TV is not available" and I really want to understand why.

Quentin-Cloos commented:
> Bearing in mind I have no idea what I'm doing and am just feeling my way around the code - this seems to work reasonably well:
> 
> ```
>     async def async_turn_off(self, **kwargs) -> None:
>         """Turn off ambilight."""
> 
>         if not self._tv.on:
>             raise Exception("TV is not available")
>             
>         if self._tv.ambilight_power == "On":
>             if not await self._tv.setAmbilightPower("Off"):
>                 raise Exception("Failed to set ambilight power")
> 
>         # current = self._tv.ambilight_current_configuration
>         # if current and current["isExpert"]:
>         #     config: AmbilightCurrentConfiguration = {
>         #         "styleName": "FOLLOW_VIDEO",
>         #         "isExpert": False,
>         #         "menuSetting": "Standard",
>         #     }
>         #     if not await self._tv.setAmbilightCurrentConfiguration(config):
>         #         raise Exception("Failed to set ambilight configuration")
>         # else:
>         #     if not await self._tv.setAmbilightMode("internal"):
>         #         raise Exception("Failed to set ambilight mode")
> 
>         self._update_from_coordinator()
>         self.async_write_ha_state()
> ```
> 
> There's a bit of a delay to update status when turning off from any mode other than Manual, but it seems to do the trick...
> 
> EDIT: sorry I haven't forked but as I say, I have no idea what I'm doing when it comes to Python so have no idea what I'm breaking...

I did that same change for fixing this issue, by the way I don't understand this part of the code and why it has to trigger the Follow_Video mode 🧐 (But I'm not a dev at all)

elupus commented:
I'll see if I can add a service for arbitrary post commands.

Quentin-Cloos commented:
> .storage - core.config_entries

Thank you! I've removed all references of my old username and password, I hope this will fix my issue 🙂

elupus commented:
I just tested the powerOff thing you did. It did not behave as wanted.
1 . TV in its in standard mode. With ambilight following what is on screen.
2. Turn off TV
3. Set an ambilight mode with screen off from home assistant
4. Turn off the ambilight light component in home assistant (it then issuing a setPowerState("Off"))
5. Turn on TV using remote control

After that sequence. Ambilight is no longer following what is on the TV anymore.

Quentin-Cloos commented:
> I just tested the powerOff thing you did. It did not behave as wanted.
> 1 . TV in its in standard mode. With ambilight following what is on screen.
> 2. Turn off TV
> 3. Set an ambilight mode with screen off from home assistant
> 4. Turn off the ambilight light component in home assistant (it then issuing a setPowerState("Off"))
> 5. Turn on TV using remote control
> 
> After that sequence. Ambilight is no longer following what is on the TV anymore.

I have tried your code again and when I turn off I have a white flash then the Ambilight change in Follow_Video mode (when the TV is turned off or turned on).

elupus commented:
Can you try. This.

Tv in off/screen off state
Turn on the light component in ha.
Switch it to any effect other than mode: manual or mode: lounge. Style: lounge should be okey.
Turn of the light entity from HA.

If my theory is correct, it should turn of properly then.

elupus commented:
I changed the code a bit. If my hunch was right. This change should fix it.

Quentin-Cloos commented:
> I changed the code a bit. If my hunch was right. This change should fix it.

I just tried your new code and I still have the exact same issue :/ I tried what you suggested but that don't change that, my Ambilight stay in Follow_Video mode when I turn it off. By the way I still have a lot of "Error executing script. Unexpected error for call_service at pos 13: TV is not available", it's really annoying and I don't had that kind of issues with the previous Ambilight component I was using (the Ambilight component from HACS), sometime it was disconnected but it didn't block my scripts, and also that component became available once again after a few seconds, since I have your component when my "TV is not available" I really need to restart my TV and HA. Don't you have that issue when you turn it off and on multiple times?

elupus commented:
I have it if i hammer it with ambilight commands. Have you turned off your other integration?

Your scrips abort because we signal an error. Just ignoring it feels wrong.

elupus commented:
Ps. It should recover after 30 seconds. That is the polling interval.

elupus commented:
So i think i've improved the behaviour where the TV ends up being marked as offline. Could you guys test?

Quentin-Cloos commented:
> Ps. It should recover after 30 seconds. That is the polling interval.

Not for me sadly :/ Maybe it's the commands I send to JointSpace, but I'm not sending a lot of commands. Here is the code for my switch that turn the Ambilight+Hue on and put the Follow_Video mode, do you think that can be the problem? For some reason I can't get the Restfull switch working with https, it's why I use a command line switch. And I'm trying your new code, for now everything is working 🙂 

```
  - platform: command_line
    switches:   
      ambihue:
         friendly_name: "Ambilight+Hue Video"
         command_on: curl -u 250c252d9a92a436ad354bb4cf9d66a7:bd9090422a8420ff0f5e0f9408136447f0b080305e3179322b4c4dd1e06a7a02 -d '{"styleName":"FOLLOW_VIDEO","isExpert":false,"menuSetting":"STANDARD"}' https://192.168.0.24:1926/6/ambilight/currentconfiguration -k --digest -v ; curl -u 250c252d9a92a436ad354bb4cf9d66a7:bd9090422a8420ff0f5e0f9408136447f0b080305e3179322b4c4dd1e06a7a02 -d '{"values":[{"value":{"Nodeid":2131230783,"Controllable":"true","Available":"true","data":{"value":"true"}}}]}' https://192.168.0.24:1926/6/menuitems/settings/update -k --digest -v
         command_off: curl -u 250c252d9a92a436ad354bb4cf9d66a7:bd9090422a8420ff0f5e0f9408136447f0b080305e3179322b4c4dd1e06a7a02 -d '{"values":[{"value":{"Nodeid":2131230783,"Controllable":"true","Available":"true","data":{"value":"false"}}}]}' https://192.168.0.24:1926/6/menuitems/settings/update -k --digest -v ; curl -u 250c252d9a92a436ad354bb4cf9d66a7:bd9090422a8420ff0f5e0f9408136447f0b080305e3179322b4c4dd1e06a7a02 -d '{"styleName":"OFF","isExpert":false}' https://192.168.0.24:1926/6/ambilight/currentconfiguration -k --digest -v
```

Quentin-Cloos commented:
Hi 🙂 I have tried your new code and I still have the same issues, I switched back to the previous code I was using and corrected some lines and now everything is working like a charm (no flash, no disconnections and with more modes), here is the code (the original is from @jomwells) if that can help you 🙂 

```
### Home Assistant Platform to integrate Phillip TVs' Ambilight as a light entity using the JointSpace API ###


import json
import string
import requests
import homeassistant.helpers.config_validation as cv
import voluptuous as vol

from homeassistant.components.light import (ATTR_BRIGHTNESS, LightEntity, PLATFORM_SCHEMA, ATTR_HS_COLOR, ATTR_TRANSITION,
                                            SUPPORT_BRIGHTNESS, SUPPORT_COLOR, SUPPORT_TRANSITION, ATTR_EFFECT, SUPPORT_EFFECT)
from homeassistant.const import (CONF_HOST, CONF_NAME, CONF_USERNAME, CONF_PASSWORD)
from requests.auth import HTTPDigestAuth
from requests.adapters import HTTPAdapter

DEFAULT_DEVICE = 'default'
DEFAULT_HOST = '127.0.0.1'
DEFAULT_USER = 'user'
DEFAULT_PASS = 'pass'
DEFAULT_NAME = 'TV Ambilights'
BASE_URL = 'https://{0}:1926/6/{1}' # for older philps tv's, try changing this to 'http://{0}:1925/1/{1}'
DEFAULT_HUE = 360
DEFAULT_SATURATION = 0
DEFAULT_BRIGHTNESS = 255
TIMEOUT = 5.0
CONNFAILCOUNT = 5


PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
	vol.Required(CONF_HOST, default=DEFAULT_HOST): cv.string,
	vol.Required(CONF_USERNAME, default=DEFAULT_USER): cv.string,
	vol.Required(CONF_PASSWORD, default=DEFAULT_PASS): cv.string,
	vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string
})

# these are the names of the effects, change the names in the quotes to change the name displayed on the front-end
EFFECT_MANUAL = "Manual"
EFFECT_STANDARD = "Standard"
EFFECT_NATURAL = "Natural"
EFFECT_IMMERSIVE = "Football"
EFFECT_VIVID = "Vivid"
EFFECT_GAME = "Game"
EFFECT_COMFORT = "Comfort"
EFFECT_RELAX = "Relax"
EFFECT_ADAP_BRIGHTNESS = "Lumina"
EFFECT_ADAP_COLOR = "Colora"
EFFECT_RETRO = "Retro"
EFFECT_SPECTRUM = "Spectrum"
EFFECT_SCANNER = "Scanner"
EFFECT_RHYTHM = "Rhythm"
EFFECT_RANDOM = "Party"
DEFAULT_EFFECT = EFFECT_MANUAL

# this is the list of effects, you can safely remove any effects from the list below to remove them from the front-end
AMBILIGHT_EFFECT_LIST = [EFFECT_MANUAL, EFFECT_STANDARD, EFFECT_NATURAL, EFFECT_IMMERSIVE, EFFECT_VIVID, 
                        EFFECT_GAME, EFFECT_COMFORT, EFFECT_RELAX, EFFECT_ADAP_BRIGHTNESS, EFFECT_ADAP_COLOR,
                        EFFECT_RETRO, EFFECT_SPECTRUM, EFFECT_SCANNER, EFFECT_RHYTHM, EFFECT_RANDOM]

def setup_platform(hass, config, add_devices, discovery_info=None):
	name = config.get(CONF_NAME)
	host = config.get(CONF_HOST)
	user = config.get(CONF_USERNAME)
	password = config.get(CONF_PASSWORD)
	add_devices([Ambilight(name, host, user, password)])

OLD_STATE = [DEFAULT_HUE, DEFAULT_SATURATION, DEFAULT_BRIGHTNESS, DEFAULT_EFFECT]

class Ambilight(LightEntity):

    def __init__(self, name, host, user, password):
        self._name = name
        self._host = host
        self._user = user
        self._password = password
        self._state = None
        self._connfail = 0
        self._brightness = None
        self._hs = None
        self._available = False
        self._effect = None
        self._session = requests.Session()
        self._session.mount('https://', HTTPAdapter(pool_connections=1))


    @property
    def name(self):
        return self._name

    @property
    def is_on(self):
        return self._state

    @property
    def available(self):
        return self._available

    @property
    def supported_features(self):
        return SUPPORT_BRIGHTNESS | SUPPORT_COLOR | SUPPORT_EFFECT

    @property
    def effect_list(self):
        return AMBILIGHT_EFFECT_LIST

    @property
    def brightness(self):
        return self._brightness

    @property
    def hs_color(self):
        return self._hs

    @property
    def effect(self):
        return self._effect

    @property
    def should_poll(self):
        return True

    def turn_on(self, **kwargs):
        if ATTR_TRANSITION in kwargs:
            # Here we save current color and brightness
            convertedHue_old = int(self._hs[0]*(255/360))
            convertedBrightness_old = self._brightness
            if ATTR_BRIGHTNESS in kwargs:
                convertedBrightness = kwargs[ATTR_BRIGHTNESS]
            else:
                convertedBrightness = self._brightness
            self._hs = kwargs[ATTR_HS_COLOR] if (ATTR_HS_COLOR in kwargs) else self._hs
            convertedHue = int(self._hs[0]*(255/360))
            convertedSaturation = int(self._hs[1]*(255/100))
            hue_addorsubst = 1 if (convertedHue_old < convertedHue) else -1
            bright_addorsubst = 1 if (convertedBrightness_old < convertedBrightness) else -1
            # Now we start transition from old color/brightness to new color and brightness. Loop until match.
            while convertedHue_old != convertedHue or convertedBrightness_old != convertedBrightness:
                convertedHue_old = (convertedHue_old + hue_addorsubst) if convertedHue_old != convertedHue else convertedHue
                convertedBrightness_old = (convertedBrightness_old + bright_addorsubst) if convertedBrightness_old != convertedBrightness else convertedBrightness
                self._postReq('ambilight/currentconfiguration',{"styleName":"Lounge light","isExpert":True,"algorithm":"MANUAL_HUE",
                "colorSettings":{"color":{"hue":convertedHue_old,"saturation":convertedSaturation,"brightness":convertedBrightness_old},
                "colorDelta":{"hue":0,"saturation":0,"brightness":0},"speed":255}} )
            self.getState()
        elif ATTR_HS_COLOR in kwargs:
            self._hs = kwargs[ATTR_HS_COLOR]
            convertedHue = int(self._hs[0]*(255/360))
            convertedSaturation = int(self._hs[1]*(255/100))
            if ATTR_BRIGHTNESS in kwargs:
                convertedBrightness = kwargs[ATTR_BRIGHTNESS]
            else:
                convertedBrightness = self._brightness
            self._postReq('ambilight/currentconfiguration',{"styleName":"Lounge light","isExpert":True,"algorithm":"MANUAL_HUE",
            "colorSettings":{"color":{"hue":convertedHue,"saturation":convertedSaturation,"brightness":convertedBrightness},
            "colorDelta":{"hue":0,"saturation":0,"brightness":0},"speed":255}} )

        elif ATTR_BRIGHTNESS in kwargs:
            # use this section instead if you cannot change the brightness without the bulb changing colour
            # (brightness commands are limited to integer values 1:10)
            #convertedBrightness = int(10*(kwargs[ATTR_BRIGHTNESS])/255)
            #self._postReq('menuitems/settings/update', "values": [{"value": {"Nodeid": 2131230769, "Controllable": "true", "Available": "true", "string_id": "Brightness", "data": {"value": convertedBrightness}}}]} )
            # and comment out all of the following
            
            convertedBrightness = kwargs[ATTR_BRIGHTNESS]
            self._postReq('ambilight/currentconfiguration',{"styleName":"Lounge light","isExpert":True,"algorithm":"MANUAL_HUE",
            "colorSettings":{"color":{"hue":int(self._hs[0]*(255/360)),"saturation":int(self._hs[1]*(255/100)),
            "brightness":convertedBrightness},"colorDelta":{"hue":0,"saturation":0,"brightness":0},"speed":255}} )

        elif ATTR_EFFECT in kwargs:
            effect = kwargs[ATTR_EFFECT]
            self.set_effect(effect)

        else:
            if OLD_STATE[3] == EFFECT_MANUAL:
                self._postReq('ambilight/currentconfiguration',{"styleName":"Lounge light","isExpert":True,"algorithm":"MANUAL_HUE",
                "colorSettings":{"color":{"hue":int(OLD_STATE[0]*(255/360)),"saturation":int(OLD_STATE[1]*(255/100)),
                "brightness":OLD_STATE[2]},"colorDelta":{"hue":0,"saturation":0,"brightness":0},"speed":255}} )
            else: 
                effect = self._effect
                self.set_effect(effect)

    def turn_off(self, **kwargs):
        global OLD_STATE
        OLD_STATE = [self._hs[0], self._hs[1], self._brightness, self._effect]
        self._postReq('ambilight/power', {'power':'Off'})
        self._state = False
		
    def getState(self):
        fullstate = self._getReq('ambilight/currentconfiguration')
        if fullstate:
            self._available = True
            styleName = fullstate['styleName']
            if styleName:
                if styleName == 'Lounge light':
                    isExpert = fullstate['isExpert']
                    if isExpert == True:
                        self._state = True
                        colorSettings = fullstate['colorSettings']
                        color = colorSettings['color']
                        hue = color['hue']
                        saturation = color['saturation']
                        bright = color['brightness']
                        self._hs = (hue*(360/255),saturation*(100/255))
                        self._brightness = bright
                        self._effect = EFFECT_MANUAL
                    else:
                        self._hs = (DEFAULT_HUE, DEFAULT_SATURATION)
                        self._brightness = DEFAULT_BRIGHTNESS

                elif styleName == 'FOLLOW_VIDEO':
                    self._state = True
                    self._hs = (DEFAULT_HUE, DEFAULT_SATURATION)
                    self._brightness = DEFAULT_BRIGHTNESS
                    effect = fullstate['menuSetting']
                    if effect == "STANDARD":
                        self._effect = EFFECT_STANDARD
                    elif effect == "NATURAL":
                        self._effect = EFFECT_NATURAL
                    elif effect == "IMMERSIVE":
                        self._effect = EFFECT_IMMERSIVE
                    elif effect == "VIVID":
                        self._effect = EFFECT_VIVID
                    elif effect == "GAME":
                        self._effect = EFFECT_GAME
                    elif effect == "COMFORT":
                        self._effect = EFFECT_COMFORT
                    elif effect == "RELAX":
                        self._effect = EFFECT_RELAX
                    
                elif styleName == 'FOLLOW_AUDIO':
                    self._state = True
                    self._hs = (DEFAULT_HUE, DEFAULT_SATURATION)
                    self._brightness = DEFAULT_BRIGHTNESS
                    effect = fullstate['menuSetting']
                    if effect == "VU_METER":
                        self._effect = EFFECT_RETRO
                    elif effect == "ENERGY_ADAPTIVE_BRIGHTNESS":
                        self._effect = EFFECT_ADAP_BRIGHTNESS
                    elif effect == "ENERGY_ADAPTIVE_COLORS":
                        self._effect = EFFECT_ADAP_COLOR  
                    elif effect == "SPECTUM_ANALYSER":
                        self._effect = EFFECT_SPECTRUM
                    elif effect == "KNIGHT_RIDER_ALTERNATING":
                        self._effect = EFFECT_SCANNER
                    elif effect == "RANDOM_PIXEL_FLASH":
                        self._effect = EFFECT_RHYTHM
                    elif effect == "MODE_RANDOM":
                        self._effect = EFFECT_RANDOM

        else:
            self._available = False
            self._state = False

    def update(self):
        self.getState()

    def set_effect(self, effect):
        if effect:
            if effect == EFFECT_MANUAL:
                self._postReq('ambilight/currentconfiguration',{"styleName":"Lounge light","isExpert":True,"algorithm":"MANUAL_HUE",
                "colorSettings":{"color":{"hue":int(OLD_STATE[0]*(255/360)),"saturation":int(OLD_STATE[1]*(255/100)),
                "brightness":OLD_STATE[2]},"colorDelta":{"hue":0,"saturation":0,"brightness":0},"speed":255}} )
                self._hs = (OLD_STATE[0], OLD_STATE[1])
                self._brightness = OLD_STATE[2]
            elif effect == EFFECT_STANDARD:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"STANDARD"})
            elif effect == EFFECT_NATURAL:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"NATURAL"})
            elif effect == EFFECT_IMMERSIVE:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"IMMERSIVE"})
            elif effect == EFFECT_VIVID:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"VIVID"})
            elif effect == EFFECT_GAME:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"GAME"})
            elif effect == EFFECT_COMFORT:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"COMFORT"})
            elif effect == EFFECT_RELAX:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_VIDEO","isExpert":False,"menuSetting":"RELAX"})
            elif effect == EFFECT_ADAP_BRIGHTNESS:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"ENERGY_ADAPTIVE_BRIGHTNESS"})
            elif effect == EFFECT_ADAP_COLOR:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"ENERGY_ADAPTIVE_COLORS"})
            elif effect == EFFECT_RETRO:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"VU_METER"})
            elif effect == EFFECT_SPECTRUM:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"SPECTRUM_ANALYSER"})
            elif effect == EFFECT_SCANNER:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"KNIGHT_RIDER_ALTERNATING"})
            elif effect == EFFECT_RHYTHM:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"RANDOM_PIXEL_FLASH"})
            elif effect == EFFECT_RANDOM:
                self._postReq('ambilight/currentconfiguration', {"styleName":"FOLLOW_AUDIO","isExpert":False,"menuSetting":"MODE_RANDOM"})

    def _getReq(self, path):
        try:
            if self._connfail:
                self._connfail -= 1
                return None
            resp = self._session.get(BASE_URL.format(self._host, path), verify=False, auth=HTTPDigestAuth(self._user, self._password), timeout=TIMEOUT)
            self.on = True
            return json.loads(resp.text)
        except requests.exceptions.RequestException as err:
            self._connfail = CONNFAILCOUNT
            self.on = False
            return None

    def _postReq(self, path, data):
        try:
            if self._connfail:
                self._connfail -= 1
                return False
            resp = self._session.post(BASE_URL.format(self._host, path), data=json.dumps(data), verify=False, auth=HTTPDigestAuth(self._user, self._password), timeout=TIMEOUT)
            self.on = True
            if resp.status_code == 200:
                return True
            else:
                return False
        except requests.exceptions.RequestException as err:
            self._connfail = CONNFAILCOUNT
            self.on = False
            return False
```

elupus commented:
Logs when it disconnect would help me. That code does not.

Quentin-Cloos commented:
> Logs when it disconnect would help me. That code does not.

Hey! Sorry, I just really want to help 🙂 I've tried to reinstall it from your last update (for sending logs) but I have this issue : 

```
Logger: homeassistant.setup
Source: setup.py:280
First occurred: 18:51:50 (1 occurrences)
Last logged: 18:51:50

Unable to prepare setup for platform philips_js.light: Platform not found (cannot import name 'COLOR_MODE_HS' from 'homeassistant.components.light' (/usr/src/homeassistant/homeassistant/components/light/__init__.py)).
```

elupus commented:
Right it's not available on older installs. So sadly that install method doesn't work anymore. I will ponder what to do in that regard.

Quentin-Cloos commented:
> Right it's not available on older installs. So sadly that install method doesn't work anymore. I will ponder what to do in that regard.

I'm updating HA right now, I was a bit scared with some breaking changes but I need to move on ahah 😛

Quentin-Cloos commented:
Hi, I have updated HA to 2021.3.4 but I still have the same issue, is there something i did not understand in your last comment?

elupus commented:
Yes..it's just dev branch now. Needs next release or beta. That is why I sorted though about how to handle it.

aegjoyce commented:
Just clean installed this latest version. Getting the following error:

```
Logger: homeassistant.setup
Source: setup.py:280 
First occurred: 21:01:36 (1 occurrences) 
Last logged: 21:01:36

Unable to prepare setup for platform philips_js.light: Platform not found (cannot import name 'COLOR_MODE_HS' from 'homeassistant.components.light' (/usr/src/homeassistant/homeassistant/components/light/__init__.py)).
```

elupus commented:
You need beta

aegjoyce commented:
> You need beta

Yeah sorry I saw and then deleted my post - apologies

aegjoyce commented:
OK, finally been able to test latest version:

In general seems to work ok, quite responsive, but lots of issues remain on my 55OLED804:

If TV is off, integration can turn on ambilight, but can't turn it off - it briefly flicks off and then turns back on at low brightness in white. Integration then reports light as being off.
It doesn't display or appear to detect the dynamic ambilight modes - just "lounge light automatic_hue", "mode: manual", "mode: expert", and "mode: lounge", all of which are static light settings
If TV is switched off in non-manual mode, then turned back on (which automatically turns on ambilight in follow_video mode), integration does not detect it being turned on
If TV is switched off whilst manual colour set, integration detects the light as being turned off, but delay of several seconds
If TV is then switched back on after previously being in manual mode, or if manual mode was used while TV was off, it detects light as being switched back on. However this is delayed by a few seconds and while the TV will turn back on into follow_video mode, the integration will report is as being in manual mode
Switching colours is responsive but can get a delay of up to 2 seconds at times
Brightness settings are weird - can be laggy (up to 2s) and anything above 30% brightness is interpreted by the tv as 100% brightness - maybe a number conversion issue?

Hope some of this is helpful. I note that every action with this light integration results in the following log:

```
Logger: haphilipsjs
Source: /usr/local/lib/python3.8/site-packages/haphilipsjs/__init__.py:110 
First occurred: 21:52:06 (124 occurrences) 
Last logged: 22:18:03

Non json data:
```

elupus commented:
@aegjoyce just one comment. All follow video type modes is considered an off state for the light entity. This is by design at the moment.

elupus commented:
@aegjoyce

* I changed it's behaviour to represent the lights even when used for video. Feels a bit more intuitive.  It now also shows some of the dynamic modes. It's a static list at the moment in the library. I suspect i can find the exact list somewhere.
* The non json data log is really no issue. It's just the TV being non json compatible. it's just a debug log.
* The laggyness i see too. I've not been able to figure out why it occurs. My suspicion is something todo with the connection pooling.
* In regards to brightness, do you see any difference in "manual" effect or "lounge: manual hue"? I think it's gamma issue.

aegjoyce commented:
Mode: lounge now makes HA think it's turned off.
Mode: expert and Mode: manual both hit max brightness at 30%.
Lounge light: MANUAL HUE: Expert and Lounge Light: AUTOMATIC HUE: Expert both hit max brightness at about 50%.
All other Lounge Light: xxxx modes mark the light as on, but don't do anything else, and don't switch to anticipated mode.
Off still doesn't work unfortunately, leaving the light in a dimmed state.

elupus commented:
"Mode: lounge" is rather pointless. It's some legacy lounge light setting. So maybe i should filter that out.

Turn off still not working is so damn annoying. I really don't want to turn off and completely disable ambilight. It seems to work on my device. 

Ok on the brightness. Must be some gamma issue.

aegjoyce commented:
> Turn off still not working is so damn annoying. I really don't want to turn off and completely disable ambilight. It seems to work on my device.

This might be the issue. So rather than a proper 'turn off' command, are you instructing ambilight to switch down to minimum or 0% brightness? Is the rationale for this the fact that if you use 'turn off' that ambilight becomes unresponsive after a while? For jomwells old custom component I use the wakelock app on the tv to prevent background sleep and this works really well. I know it's not the most elegant solution for an integration to have to load other software onto the tv but I've yet to hear of a better solution

elupus commented:
if the tv screen is off, I ask it to go back to FOLLOW_VIDEO
If the tv screen is on, i instruct it to go to OFF.

Rational being. If you use it as a lounge light with tv screen off, when you turn it off, you want the tv to still follow video again next time you actually want to watch something. Otherwise, the ambilight would be disabled next time you want to watch something.

If you are watching something, you are actually asking the ambilight to be disabled.

aegjoyce commented:
Isn't this something that could be fixed in media_player.py? Something like - if media player goes from off to on, and has ambilight functions, then fire an ambilight on + follow_video command?

aegjoyce commented:
I should add that my tv has the same behaviour as yours in this regard, I've just got around it with an automation

aegjoyce commented:
Just tested whether its the wakelock app that causes the ambilight not to turn off when switched to follow_video on a blank screen - it isn't. Also my tv becomes unresponsive within seconds after being turned off if I disable wakelock so it looks like it's almost a prerequisite for this integration.

Of note, all the follow_video modes cause the ambilight to be dimmed when the tv is off, but to different extents. Follow_video 'natural' seems to be the dimmest (tested with the jomwells component)

aegjoyce commented:
Giving this further thought, I really think the preferred action should be to separate turning on the TV from turning on the ambilight to a preset mode. Setting up an automation to set the ambilight mode when turning on the TV is very easy and shouldn't be problematic for most users. Having it hardcoded that the ambilight gets set to Standard Follow Video mode could be frustrating for many users - some people might not use Follow Video modes at all, some might use a different setting (Vivid, Natural etc.) and will have to change this manually (or by automation) each time they turn the TV on.

There's not a huge difference in the pros and cons between each method, but if hardcoding a return to Follow Video on turning off ambilight means that the ambilight doesn't turn off at all (on certain TVs) then I think the preference has to be to just do a hard 'turn off'.

elupus commented:
I think most people have it to follow video. Don't see any reason to buy a philips otherwise ;). It would be rather annoying to have it switch of just since it turned  off lounge lights. 

It would be equally easy to add an automation to change the mode after turn on to something else.

That said, we could probably remember the previous mode and restore that, if only it would behave and actually turn of completely.... I might have to resign om this due to tv bugs.

chillkiller commented:
> 
> 
> I think most people have it to follow video. Don't see any reason to buy a philips otherwise ;). It would be rather annoying to have it switch of just since it turned off lounge lights.
> 
> It would be equally easy to add an automation to change the mode after turn on to something else.
> 
> That said, we could probably remember the previous mode and restore that, if only it would behave and actually turn of completely.... I might have to resign om this due to tv bugs.

Hi, can u also integrate: https://github.com/jomwells/ambihue

??
Would be cool to have it aio
 I realy need the ambilight+hue switch for temporarily stop the colorchanges while cooking and watching TV ;-)

aegjoyce commented:
> I think most people have it to follow video. Don't see any reason to buy a philips otherwise ;). It would be rather annoying to have it switch of just since it turned off lounge lights.

Agreed, I have mine on follow video, but I set mine to Vivid - so I would still need an automation

> It would be equally easy to add an automation to change the mode after turn on to something else.

True; but for those that don't have it set to follow video (for whatever reason) this would result in the ambilight turning on first, and then turning off via automation, which could look odd. I agree this might be a rare case use, but still...

> That said, we could probably remember the previous mode and restore that

I think this would be a good way forward - some sort of config storage for the last follow video mode, and have this activate when the TV turns on.

Is it possible to link the turn on action in media_player.py to the ambilight turn on function?

> if only it would behave and actually turn of completely....

I don't think it's a bug. I'm pretty confident if you remove the option for going to follow video in the turn off command then you'll find it turns off ambilight regardless of tv model. I also think that as a result this is the ONLY option for the turn off command.

I personally think the ability to turn off the light reliably is much more important than having it restore a follow video mode when the TV is switched back on!

>  I might have to resign om this due to tv bugs.

Please don't! You've done amazing work with this so far :)

chillkiller commented:
> I don't think it's a bug. I'm pretty confident if you remove the option for going to follow video in the turn off command then you'll find it turns off ambilight regardless of tv model. I also think that as a result this is the ONLY option for the turn off command.
> 

I havent read the full discusion, so what "bug"??
can I reproduce it and how to check what you mean?

As I've fixed my issues (login reset after tv-update, yeah stupid...) all my automations work well like before and I use 3 components for my philips tv (2 custom components from @jomwells - ambihue switch & ambilights light/switch and this integration) and they all seems to be ok and do what I want from em without any conficts to each other.

But I must say I couldnt test this pr and cant say whats the different to my solution.
I just wish that all of these solutions could be combined into one integration, just like @jomwells started trying, before he stopped developing his components (dont know why).

I'm no coder/programmer, I compare the different files and make a sense of it.
So where or what is/was the problem of integrating all possible, functioning features of philips tv, Jointspace and above all ambilight+Hue switch and ambilight light/colorpicker/switch, apart from the private time expenditure, which of course I give very high credit?

Sorry, but I'm a little confused right now...

chillkiller commented:
@elupus 


>    It would be equally easy to add an automation to change the mode after turn on to something else.

>True; but for those that don't have it set to follow video (for whatever reason) this would result in the ambilight turning on first, >and then turning off via automation, which could look odd. I agree this might be a rare case use, but still...

In my opinion an integration isnt an automation, its part of an automation.
the tv itself truely have its own automations, some makes sence, other not.

I have to agree with @aegjoyce its easiy to make an automation to set the tv automatically in follow video, but if it is already coded in the integration and u dont want/need it, it will be impossible to suppress it ;)

We are working/playing with home automations, shouldn't it be up to the user what he would prefer and then automate it himself?

It's not LEGO (even if it's partly just as simple), from my point of view it would be more important to provide all possible functions and leave it up to the user what he makes of them. Sorry, I don't mean that, it's just my opinion on it.

You are discussing building automation firmly into an integration. The "restore last state" function could also be overwritten with automation and a little polling test for less delay.

As such, personally (if I could) I would steer clear of it and leave it to the users, and instead make an effort to ensure that the integration is clean and complete.

Please don't take this personally, my drive is only to help you find a good solution ...

chillkiller commented:
>    I might have to resign om this due to tv bugs.
>
>Please don't! You've done amazing work with this so far :)

I also agree and if I can help you, I'll be happy to do it !!

elupus committed:
Restore ambilight mode on delay

elupus commented:
I finally managed to understand why i wasn't able to restore ambilight mode to follow video properly. To restore, one need to turn it to off state, wait for a transition effect to off, then turn it to follow video again.

If you are too quick. The off transition gets aborted half way, and get stuck.

So.. it should now turn off properly and switch to follow video. I will try to implement a restoring off mode later.

Also, I've reduced TV hammering of ambilight data, that seemed to crash it. I will backport that.

aegjoyce commented:
Sorry, switch off still not working on OLED804. Fades off, then after a second does a little pulse, then a few seconds later comes on again at dim brightness...

elupus committed:
Drop the ability to restore to follow video

The tv's do invalid things and start showing
background color when asked to follow video
while off.

elupus committed:
2.7.4

aegjoyce commented:
This looks good! Turns on and off correctly, and the other tweaks seem to make it really responsive. I'm impressed.

Any particular things you want testing? I think the only thing I would look to add would be some sort of lookup dictionary so that the colour modes are labelled in a friendlier way (matching up with the built-in names on the TV) and without some of the duplicate modes. 

Functionally though it's excellent!

balloob commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,411 @@
+    @property
+    def supported_features(self):
+        """Return supported features on this light."""
+        if self.color_mode == COLOR_MODE_HS:
```
Supported features should be static.

balloob commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,411 @@
+    def effect_list(self):
+        """Return the list of supported effects."""
+        effects = []
+        effects.extend(
```
All this seems pretty expensive to calculate, is this something we can cache ?

elupus replied:
Maybe. It depends on state. But mainly it depends on if the TV is on or not. So maybe i can have two lists.

elupus replied:
I added cache so it keeps track of any changes.

balloob commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,411 @@
+        return SUPPORT_EFFECT
+
+    @property
+    def should_poll(self):
```
This is the default in the coordinator entity that you inherit from.

balloob commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,411 @@
+        self._tv = coordinator.api
+        self._hs = None
+        self._brightness = None
+        self._system = system
+        self._unique_id = unique_id
+        self._coordinator = coordinator
+        super().__init__(coordinator)
+
+        self._update_from_coordinator()
+
+    @property
+    def name(self):
+        """Return the device name."""
+        return self._system["name"]
+
+    @property
+    def unique_id(self):
+        """Return unique identifier if known."""
+        return self._unique_id
```
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity.py#L171
https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py#L603-L616

```suggestion
        self._system = system
        self._attr_name = system["name"]
        self._attr_unique_id = unique_id
        self._coordinator = coordinator
        super().__init__(coordinator)

        self._update_from_coordinator()
```

balloob commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,411 @@
+
+        return _get_effect(EFFECT_MODE, self._tv.ambilight_mode, None)
+
+    @property
+    def supported_color_modes(self):
+        """Return the supported color mode."""
+        return [COLOR_MODE_HS, COLOR_MODE_ONOFF]
+
```
Replace by setting `self._supported_color_modes` in constructor. 

https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py#L603-L616

balloob commented on homeassistant/components/philips_js/light.py:
```diff
@@ -0,0 +1,411 @@
+                color["saturation"] * 100.0 / 255.0,
+            )
+            self._brightness = color["brightness"]
+        else:
+            data = self._tv.ambilight_cached
+            if data:
+                hsv_h, hsv_s, hsv_v = color_RGB_to_hsv(*_average_pixels(data))
+                self._hs = hsv_h, hsv_s
+                self._brightness = hsv_v * 255.0 / 100.0
+            else:
+                self._hsv = None
+                self._brightness = None
```
```suggestion
        elif data := self._tv.ambilight_cached:
            hsv_h, hsv_s, hsv_v = color_RGB_to_hsv(*_average_pixels(data))
            self._hs = hsv_h, hsv_s
            self._brightness = hsv_v * 255.0 / 100.0
        else:
            self._hsv = None
            self._hs = None
            self._brightness = None
```

balloob reviewed (approved):
Looks good! Left a couple of comments.

elupus committed:
Use new _attr_* variables for constants

elupus committed:
Simplify if else clause

elupus committed:
Calculate effects on update instead of on the fly

elupus committed:
No need to set the should_poll

It is inherited from CoordinatorEntity

elupus committed:
Only update effect when data changes

aegjoyce commented:
These latest changes have broken something. Re-installed the component from scratch using the new commits, it can't add the light entity:

```
Logger: homeassistant.components.light
Source: custom_components/philips_js/light.py:62 
Integration: Light (documentation, issues) 
First occurred: 12:25:17 (2 occurrences) 
Last logged: 12:25:17

Error adding entities for domain light with platform philips_js
Error while setting up philips_js platform for light
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 364, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 558, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 581, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 334, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 368, in _async_write_ha_state
    state = self._stringify_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 340, in _stringify_state
    state = self.state
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 738, in state
    return STATE_ON if self.is_on else STATE_OFF
  File "/config/custom_components/philips_js/light.py", line 232, in is_on
    mode, style, _ = _parse_effect(self.effect)
  File "/config/custom_components/philips_js/light.py", line 62, in _parse_effect
    style, _, algorithm = effect.partition(EFFECT_PARTITION)
AttributeError: 'NoneType' object has no attribute 'partition'
```

aegjoyce commented:
Can confirm that reverting back to 31c71d2 fixes the issue.

elupus commented:
I missed something in the upgrade to attributes. Will fix later.

elupus commented:
Oh right. you need lasted dev branch for those things to work properly. It's not fault in the pull request. Just won't work with older home assistant versions.
